### PR TITLE
Use cleaner user visible URLs via pushState()

### PR DESCRIPTION
**Pull request by stefwalter**
Use cleaner URLs via pushState() in shell index
    
 * We produce clean and bookmarkable URLs now.
 * The hash is passed directly to the component.
 * The @localhost doesn't show up in address bar.
 * When switching between servers we track component by label.

We serve the shell HTML under most paths requested from cockpit. The current rules are:
    
 * Don't serve under paths starting with /cockpit
 * Serve under paths starting with /@host
 * Serve under paths starting with valid package names

### CODE REVIEW

**Comment by mvollmer** [comment]
On first reading, I thought that "Serve under paths starting with valid package names" means only to serve when the path starts with the name of an actually existing package.  In truth, we serve the shell when the path is syntactically a valid package name, and that excludes things like favicon.ico etc.

File: pkg/dashboard/manifest.json [hunk 1]
```diff
@@ -2,12 +2,9 @@
...
     "dashboard": {
-        "list": {
-            "path": "dashboard/list",
+        "index": {
             "label": "Dashboard",
             "order": 0
         }
-    },
-
-    "limit": 5
```
Comment (mvollmer) [inline pkg/dashboard/manifest.json]: Is it a mistake to remove the "limit" here?

**Comment by mvollmer** [comment]
`pkg/users/manifest.json` needs the local -> index change as well, I think.

File: pkg/dashboard/manifest.json (hunk 1 above)
Comment (mvollmer) [inline pkg/dashboard/manifest.json]: Looks like this field is unused.

**Comment by mvollmer** [comment]
Navigating with the forward and back arrow buttons of the browser seems inconsistent.  (I haven't checked this earlier, so maybe this isn't a regression...)

Navigating by clicking somehwere in our navbar, for example, seems to put two entries in the history.

- Open a new tab and load Cockpit into it.
- Click on "Services"
- Click back arrow
- Still on "Services"
- Clock back arrow again
- Now on "System"

**Comment by stefwalter** [comment]
> Navigating by clicking somehwere in our navbar, for example, seems to put two entries in the history.

I fixed most cases of this. Some cases (such as with angularjs in the kubernetes bar) still occur. This is because of cases where the ```#hash``` changes, sometimes to an empty ```#/```

File: pkg/shell/index.js [hunk 2]
```diff
@@ -378,7 +490,6 @@ define([
             label = _("Machines");
         }
         $("#machine-link span").text(label);
-        $("title").text(title || default_title);
```
Comment (petervo) [inline pkg/shell/index.js]: I think this was the only use of title. Since it's removed should we remove the declaration and assignment above as well?

Comment (stefwalter) [inline pkg/shell/index.js]: Good catch. Done.

**Comment by petervo** [comment]
When an old server is a primary server, the links on the dashboard explicitly go to system/host. I think we need a redirect a for that.

**Comment by stefwalter** [comment]
> When an old server is a primary server, the links on the dashboard explicitly go to system/host. I think we need a redirect a for that.

This is hard to do, both for the dashboard, and for the defaults (when running on an old index.html), because the old code expicitly maps in shell/shell.html for this case. At a very minimum the components are accessible by expcitly clicking around.

In any case can we fix this in a follow up, and later commit?

**Comment by petervo** [comment]
I'm ok with waiting, it's just that it's easy to think that server isn't working at all when the first thing you see looks broken, esp when you have a version that doesn't have the nice 404 page.

**Pull request closed without merging**